cliquet/storage/postgresql/__init__.py

```diff
@@ -0,0 +1,458 @@
+
+        if limit:
+            assert isinstance(limit, six.integer_types)  # validated in view
+            safeholders['pagination_limit'] = 'LIMIT %s' % limit
+
+        with self.connect() as cursor:
+            cursor.execute(query % safeholders, placeholders)
+            results = cursor.fetchmany(self._max_fetch_size)
+
+        if not len(results):
+            return [], 0
+
+        count_total = results[0]['count_total']
+
+        records = []
+        for result in results:
+            record = result['data']
+            record[resource.id_field] = result['id']
+            record[resource.modified_field] = result['last_modified']
+            records.append(record)
+
+        return records, count_total
+
+    def _format_conditions(self, resource, filters, prefix='filters'):
+        """Format the filters list in SQL, with placeholders for safe escaping.
+
+        :note:
+            All conditions are combined using AND.
+
+        :note:
+            Field name and value are escaped as they come from HTTP API.
+
+        :returns: A SQL string with placeholders, and a dict mapping
+            placeholders to actual values.
+        :rtype: tuple
+        """
+        operators = {
+            COMPARISON.EQ: '=',
+            COMPARISON.NOT: '<>',
+        }
+
+        conditions = []
+        holders = {}
+        for i, filter_ in enumerate(filters):
+            field, value, operator = filter_
+
+            if field == resource.id_field:
+                sql_field = 'id'
+            elif field == resource.modified_field:
+                sql_field = 'last_modified::BIGINT'
+            else:
+                # Safely escape field name
+                field_holder = '%s_field_%s' % (prefix, i)
+                holders[field_holder] = field
+                # JSON operator ->> retrieves values as text.
+                # If field is missing, we default to ''.
+                sql_field = "coalesce(data->>%%(%s)s, '')" % field_holder
+                # JSON-ify the native value (e.g. True -> 'true')
+                value = json.dumps(value).strip('"')
+
+            # Safely escape value
+            value_holder = '%s_value_%s' % (prefix, i)
+            holders[value_holder] = value
+
+            sql_operator = operators.setdefault(operator, operator)
+            cond = "%s %s %%(%s)s" % (sql_field, sql_operator, value_holder)
+            conditions.append(cond)
+
+        safe_sql = ' AND '.join(conditions)
+        return safe_sql, holders
+
+    def _format_pagination(self, resource, pagination_rules):
+        """Format the pagination rules in SQL, with placeholders for
+        safe escaping.
+
+        :note:
+            All rules are combined using OR.
+
+        :note:
+            Field names are escaped as they come from HTTP API.
+
+        :returns: A SQL string with placeholders, and a dict mapping
+            placeholders to actual values.
+        :rtype: tuple
```

Comment:
Interesting, it's the first time I see this, we should be consistent and put if everywhere if we start to.